ipc/provider/src/manager/evm/dryrun.rs

```diff
@@ -0,0 +1,405 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: MIT
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use ethers::abi::{Function, FunctionExt, Tokenizable};
+use ethers::types::{Selector, U256};
+use ethers_contract::encode_function_data;
+use ipc_actors_abis::register_subnet_facet;
+use ipc_api::evm::payload_to_evm_address;
+use std::collections::btree_map::BTreeMap;
+use std::collections::HashMap;
+
+use crate::config::serialize::{serialize_address_to_str, serialize_bytes_to_str};
+use crate::manager::{
+    BottomUpCheckpointRelayer, EthSubnetManager, GetBlockHashResult, SubnetGenesisInfo, SubnetInfo,
+    SubnetManager, TopDownFinalityQuery, TopDownQueryPayload,
+};
+use crate::VMType;
+use fvm_ipld_encoding::{BytesSer, RawBytes};
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::MethodNum;
+use ipc_api::checkpoint::{
+    BottomUpCheckpoint, BottomUpCheckpointBundle, QuorumReachedEvent, Signature,
+};
+use ipc_api::cross::IpcEnvelope;
+use ipc_api::eth_to_fil_amount;
+use ipc_api::ethers_address_to_fil_address;
+use ipc_api::staking::{StakingChangeRequest, ValidatorInfo};
+use ipc_api::subnet::ConstructParams;
+use ipc_api::subnet_id::SubnetID;
+use serde::Serialize;
+
+const INVOKE_CONTRACT: MethodNum = 3844450837;
+
+#[derive(Serialize)]
+struct MockedFVMTxn {
+    #[serde(serialize_with = "serialize_address_to_str")]
+    from: Address,
+    #[serde(serialize_with = "serialize_address_to_str")]
+    to: Address,
+    /// The value, display as string to align with evm display
+    value: String,
+    #[serde(serialize_with = "serialize_bytes_to_str")]
+    method_params: Vec<u8>,
+    method: MethodNum,
+}
+
+#[derive(Serialize)]
+struct MockedEVMTxn {
+    from: ethers::types::Address,
+    to: ethers::types::Address,
+    value: U256,
+    #[serde(serialize_with = "serialize_bytes_to_str")]
+    calldata: Vec<u8>,
+    #[serde(serialize_with = "serialize_bytes_to_str")]
+    method: Selector,
```

Comment:
Can you add a docstring about the relationship between `method` and `calldata`? I see `method` is not used in the conversion below.